core/os/thread.cpp

```diff
@@ -50,8 +50,8 @@ void Thread::_set_platform_functions(const PlatformFunctions &p_functions) {
 	platform_functions = p_functions;
 }
 
-void Thread::callback(Thread *p_self, const Settings &p_settings, Callback p_callback, void *p_userdata) {
-	Thread::caller_id = _thread_id_hash(p_self->thread.get_id());
+void Thread::callback(ID p_caller_id, const Settings &p_settings, Callback p_callback, void *p_userdata) {
```

Comment:
I'm curious about why this is needed. I mean, if the pointer was only needed for getting the id, this new version seems to be safer. However, in practice `Thread` object shouldn't be destroyed before the thread ends.

Reply:
There's a call to `thread.swap(new_thread)` right after `new_thread` gets created, so in the old version, it's ambiguous whether `p_self->thread` refers to the old `thread` or the just created `new_thread`.

Threading is hard :smiling_face_with_tear:

Reply:
Besides `swap`, `Thread` technically can be destroyed (and will detach thread in this case) before the thread is finished. So it is safer.